src/main/java/org/prebid/server/auction/ExchangeService.java

```diff
@@ -419,43 +457,40 @@ private boolean isValidBidder(String bidder, BidderAliases aliases) {
      * <p>
-     * Also, removes user.keywords, gender, yob, geo and ext (except user.ext.eids and user.ext.digitrust)
-     * in case bidder does not use first party data.
+     * Also, removes user.ext.prebid (if present) and user.ext.data (in case bidder does not use first party data).
      */
-    private User prepareUser(User user, ExtUser extUser, String bidder, BidderAliases aliases,
-                             Map<String, String> uidsBody, UidsCookie uidsCookie, boolean useFirstPartyData) {
+    private User prepareUser(User user,
+                             ExtUser extUser,
+                             String bidder,
+                             BidderAliases aliases,
+                             Map<String, String> uidsBody,
+                             UidsCookie uidsCookie,
+                             boolean useFirstPartyData,
+                             ExtBidderConfigFpd fpdConfig) {
+
         final String updatedBuyerUid = updateUserBuyerUid(user, bidder, aliases, uidsBody, uidsCookie);
-        final boolean shouldUpdateUserExt = extUser != null && extUser.getPrebid() != null;
-        final boolean shouldRemoveUserFields = !useFirstPartyData && checkUserFieldsHavingValue(user);
+        final boolean shouldUpdateUserExt = extUser != null
+                && (extUser.getPrebid() != null || (extUser.getData() != null && !useFirstPartyData));
 
-        if (updatedBuyerUid != null || shouldRemoveUserFields || shouldUpdateUserExt) {
+        User maskedUser = user;
+        if (updatedBuyerUid != null || shouldUpdateUserExt) {
             final User.UserBuilder userBuilder = user == null ? User.builder() : user.toBuilder();
             if (updatedBuyerUid != null) {
                 userBuilder.buyeruid(updatedBuyerUid);
             }
 
-            if (shouldRemoveUserFields) {
-                final ExtUser updatedExtUser = extUser == null
-                        ? null
-                        : ExtUser.builder()
-                        .eids(extUser.getEids())
-                        .digitrust(extUser.getDigitrust())
-                        .build();
-
-                userBuilder
-                        .keywords(null)
-                        .gender(null)
-                        .yob(null)
-                        .geo(null)
-                        .ext(updatedExtUser);
-            } else if (shouldUpdateUserExt) {
-                userBuilder.ext(extUser.toBuilder().prebid(null).build());
+            if (shouldUpdateUserExt) {
+                userBuilder.ext(extUser.toBuilder()
+                        .prebid(null)
+                        .data(null)
+                        .build());
             }
 
-            return userBuilder.build();
+            maskedUser = userBuilder.build();
         }
 
-        return user;
+        final User fpdUser = fpdConfig == null ? null : fpdConfig.getUser();
+        return jsonMergeUtil.merge(fpdUser, maskedUser, User.class);
```

Comment:
I am not sure if it is possible that pfdUser from bidderConfig can contain not fpd fields, but simple merge will override non fpd fields, that is not expected behavior.  I propose to replace only fpd fields  one by one.

Reply:
Addressed this concern by switching to `FprResolver`